presto-main/src/main/java/com/facebook/presto/memory/QueryContext.java

```diff
@@ -306,36 +337,49 @@ public QueryId getQueryId()
         return queryId;
     }
 
-    public synchronized void setMemoryLimits(DataSize queryMaxTaskMemory, DataSize queryMaxTotalTaskMemory)
+    public synchronized void setMemoryLimits(DataSize queryMaxTaskMemory, DataSize queryMaxTotalTaskMemory, DataSize queryMaxBroadcastMemory)
     {
         // Don't allow session properties to increase memory beyond configured limits
         maxUserMemory = Math.min(maxUserMemory, queryMaxTaskMemory.toBytes());
         maxTotalMemory = Math.min(maxTotalMemory, queryMaxTotalTaskMemory.toBytes());
+        maxBroadcastUsedMemory = Math.min(maxBroadcastUsedMemory, queryMaxBroadcastMemory.toBytes());
     }
 
     private static class QueryMemoryReservationHandler
             implements MemoryReservationHandler
     {
         private final BiFunction<String, Long, ListenableFuture<?>> reserveMemoryFunction;
         private final BiPredicate<String, Long> tryReserveMemoryFunction;
+        private final Consumer<Long> updateBroadcastMemoryFunction;
+        private final Predicate<Long> tryUpdateBroadcastMemoryFunction;
 
         public QueryMemoryReservationHandler(
                 BiFunction<String, Long, ListenableFuture<?>> reserveMemoryFunction,
-                BiPredicate<String, Long> tryReserveMemoryFunction)
+                BiPredicate<String, Long> tryReserveMemoryFunction,
+                Consumer<Long> updateBroadcastMemoryFunction,
+                Predicate<Long> tryUpdateBroadcastMemoryFunction)
         {
             this.reserveMemoryFunction = requireNonNull(reserveMemoryFunction, "reserveMemoryFunction is null");
             this.tryReserveMemoryFunction = requireNonNull(tryReserveMemoryFunction, "tryReserveMemoryFunction is null");
+            this.updateBroadcastMemoryFunction = requireNonNull(updateBroadcastMemoryFunction, "updateBroadcastMemoryFunction is null");
+            this.tryUpdateBroadcastMemoryFunction = requireNonNull(tryUpdateBroadcastMemoryFunction, "tryUpdateBroadcastMemoryFunction is null");
         }
 
         @Override
-        public ListenableFuture<?> reserveMemory(String allocationTag, long delta)
+        public ListenableFuture<?> reserveMemory(String allocationTag, long delta, boolean enforceBroadcastMemoryLimit)
         {
+            if (enforceBroadcastMemoryLimit) {
+                updateBroadcastMemoryFunction.accept(delta);
+            }
             return reserveMemoryFunction.apply(allocationTag, delta);
         }
 
         @Override
-        public boolean tryReserveMemory(String allocationTag, long delta)
+        public boolean tryReserveMemory(String allocationTag, long delta, boolean enforceBroadcastMemoryLimit)
         {
+            if (enforceBroadcastMemoryLimit && !tryUpdateBroadcastMemoryFunction.test(delta)) {
```

Comment:
put this after tryReserveMemoryFunctinon.test()... so that local memory limit errors will be thrown first